implemented improvements to RPC history

evgeniuz opened the pull request:
The problem here was that when non-RPC provider was selected (e. g. mainnet) `rpcTarget` still contains most recently used RPC URL, so `renderCommonRpc` was hiding it (as it should be displayed anyway in `renderCustomOption`). I've fixed this by not only checking if `rpc` equals current `rpcTarget`, but selected provider is actually `rpc`.

Another issue was order: it was showing them in following order:

1. current RPC
2. oldest RPC
3. recent RPC

Now the order is following:

1. current RPC
2. recent RPC
3. oldest RPC

Which makes more sense to me. Otherwise behavior is the same (oldest one is pushed away when new one is introduced, just as it was before).

I've also increased history size from 2 to 3 (this is opinionated from discussion, please, say if I should revert this back to 2).

I wasn't sure if I need to add tests here (as this functionality wasn't tested before), so added e2e test just in case.

Fixes #4498

evgeniuz committed:
implemented improvements to RPC history

evgeniuz committed:
backported RPC improvements to oldui

bdresser commented:
thanks @evgeniuz! history size of 3 sounds good to me.

@whymarrh or @danjm could you take a look?

danjm commented on app/scripts/controllers/preferences.js:
```diff
@@ -338,7 +338,7 @@ class PreferencesController {
     if (_url !== 'http://localhost:8545') {
       rpcList.push(_url)
     }
-    if (rpcList.length > 2) {
```
Could you update the documentation for this method to reflect the change in the max length of the rpcList?

danjm commented on ui/app/components/dropdowns/network-dropdown.js:
```diff
@@ -291,17 +292,17 @@ NetworkDropdown.prototype.renderCommonRpc = function (rpcList, provider) {
         ]
       )
     }
-  })
```
This is a minor issue, but I think it will be helpful to others trying to understand the code: could you make the reversal of the list really explicit by removing the reverse here and doing
```
  const reversedRpcList = rpcList.slice(0).reverse()

  return reversedRpcList.map((rpc) => {
```
at top?

We should be better ordering these on the backend in my opinion. But that is definitely beyond the scope of this task. For now, it would at least be good to make it clear to the reader of the code what is happening.

danjm reviewed (dismissed):
I am approving this with two small suggestions for improvement. These are non-blocking suggestions and this PR can be merged without those changes, but it would be nice if they were made :)

Nice PR, and super awesome to see e2e tests.

Can you merge/rebase to get the latest changes from `develop` in here so that we see the e2e tests passing? Then we are ready to merge.

whymarrh commented:
> Nice PR, and super awesome to see e2e tests.

Agree, thanks @evgeniuz, this is awesome work!

whymarrh commented:
Do you mind rebasing this on the latest develop to include the latest e2e and CI fixes and once we have a green build we can merge this?

evgeniuz committed:
updated docs, small improvement of recent RPC rendering

evgeniuz commented:
Added suggested improvements, rebased on top of fresh `develop` branch. Don't think oldui e2e test fail for Chrome has something to do with this changeset.

whymarrh commented:
> Don't think oldui e2e test fail for Chrome has something to do with this changeset.

No worries, I gave it an extra kick and it's all good now.

danjm commented:
Thanks again, merging.
